Source/WebCore/html/HTMLAttachmentElement.cpp

```diff
@@ -70,14 +75,122 @@ HTMLAttachmentElement::~HTMLAttachmentElement() = default;
 
 Ref<HTMLAttachmentElement> HTMLAttachmentElement::create(const QualifiedName& tagName, Document& document)
 {
-    return adoptRef(*new HTMLAttachmentElement(tagName, document));
+    Ref<HTMLAttachmentElement> attachment = adoptRef(*new HTMLAttachmentElement(tagName, document));
+    if (DeprecatedGlobalSettings::attachmentWideLayoutEnabled() && DeprecatedGlobalSettings::attachmentElementEnabled()) {
+        ASSERT(attachment->m_implementation == Implementation::Legacy);
+        ASSERT(!attachment->renderer()); // Switch to modern style *must* be done before renderer is created!
+        attachment->m_implementation = Implementation::Modern;
+        if (auto shadow = attachment->userAgentShadowRoot())
+            attachment->ensureModernShadowTree(*shadow);
+        else
+            attachment->ensureUserAgentShadowRoot();
+    }
+    return attachment;
+}
+
+void HTMLAttachmentElement::didAddUserAgentShadowRoot(ShadowRoot& root)
+{
+    if (m_implementation == Implementation::Modern)
+        ensureModernShadowTree(root);
+}
+
+void HTMLAttachmentElement::ensureModernShadowTree(ShadowRoot& root)
+{
+    ASSERT(m_implementation == Implementation::Modern);
+    if (m_elementWithTitle)
+        return;
+
+    constexpr auto styleSheet = R"CSS(
+attachment::-webkit-attachment-container {
+    display: grid;
+    max-width: 500px;
+    background-color: lightgray;
+    border-radius: 8px;
+    font-family: "Lucida Grande";
+    pointer-events: none;
+    user-select: none;
+    -webkit-user-select: none;
+}
+attachment::selection {
+    background-color: blue;
+}
+attachment::-webkit-attachment-preview {
+    grid-row: 1 / 4;
+    grid-column: 1;
+    align-self: center;
+}
+attachment::-webkit-attachment-info {
+    grid-row: 1;
+    grid-column: 2;
+    font-size: x-small;
+    color: gray;
+}
+attachment::-webkit-attachment-title {
+    grid-row: 2;
+    grid-column: 2;
+    font-size: small;
+}
+attachment::-webkit-attachment-subtitle {
+    grid-row: 3;
+    grid-column: 2;
+    font-size: x-small;
+    color: rgb(82, 145, 214);
+}
+    )CSS"_s;
+
+    static MainThreadNeverDestroyed<const String> shadowStyle(StringImpl::createWithoutCopying(styleSheet.characters(), styleSheet.length()));
+
+    auto style = HTMLStyleElement::create(HTMLNames::styleTag, document(), false);
+    style->setTextContent(String { shadowStyle });
+    root.appendChild(WTFMove(style));
+
+    auto container = HTMLDivElement::create(document());
+    container->setPseudo(ShadowPseudoIds::webkitAttachmentContainer());
+    root.appendChild(container);
+
+    m_innerLegacyAttachment = adoptRef(*new HTMLAttachmentElement(HTMLNames::attachmentTag, document()));
```

Comment:
I don't feel too great about nested attachment elements as a long term approach. Do you have a plan to address this / alternatives considered? I know there was some discussion about using blob URLs for the icon.

If we are going to continue with the icon as a distinct element, this should be it's own internal element type (similar to `SpinButtonElement`)

Reply:
There are a few reasons I went this way:
- Using blob URLs was not easy (to me at least 😅), and also it felt it would have been quite expensive in both CPU and memory to perform multiple counter-acting conversions.
- We can keep the legacy code around while transitioning, so even a new WebKit library could still work with an older Mail app.
- The legacy attachment code does more than just display an image, it also handles requesting that icon/thumbnail when first rendering, showing an optional progress bar, and other operations, which would make it much more difficult to port to our "modern" element all at once.
- And it's not meant to stay long term, as eventually the legacy code should be removed once we only need the modern style and everything works with it as needed.

Reply:
Sounds reasonable to me. Can you add a FIXME + bug, so that we consider improving this in the longer term.